18F/identity-idp
# Subset fonts to used character data

## 🛠 Summary of changes

Optimizes fonts to remove unused character data.

This is similar to #6094, but more aggressively-optimizes font files. With #6094, we used predefined Latin character set to reduce the size of fonts. With this approach, we specifically target only the character glyphs that exist in localized string data.

This also removes unused `.woff` files. As of [USWDS v3.4.0](https://github.com/uswds/uswds/releases/tag/v3.4.0), only `.woff2` files are used.

For future consideration: We could optimize this even more aggressively if we created per-locale font subsets, to avoid loading French and Spanish character data for English users (and vice-versa, as applicable).

**Why?**

After `application.css`, these fonts are the second-largest first-party asset on https://secure.login.gov , each coming in around 21.5kb (vs. `22.5kb` for `application.css`), for a combined total of 42.9kb font data on the homepage.

**Performance Impact**

The numbers below reflect the combined total of `PublicSans-Regular.woff2` and `PublicSans-Bold.woff2`, the most common font combination loaded on each page (and [preloaded by default](https://github.com/18F/identity-idp/blob/31adc49c733fac90cf7048882862b98ce3a4e007/app/views/layouts/base.html.erb#L19-L20)).

Before: 42.9kb (21.5kb + 21.4kb)
After: 30kb (15kb + 15kb)
Diff: -12.9kb (-30%)

## 📜 Testing Plan

Verify that text is still rendered as Public Sans:

1. Go to http://localhost:3000 or any page of the application
2. Inspect some text using your browser's developer tools
3. Observe that the rendered font is using the Public Sans network resource
   - In Chrome, this is in the bottom "Rendered Fonts" section under ["Computed" styles](https://developer.chrome.com/docs/devtools/css/reference#computed)
